[BEAM-2193] implement join

xumingming opened the pull request:
implement join

aaltay commented:
R: @XuMingmin @takidau

takidau commented on dsls/pom.xml:
```diff
@@ -22,7 +22,7 @@
     <groupId>org.apache.beam</groupId>
     <artifactId>beam-parent</artifactId>
-    <version>0.7.0-SNAPSHOT</version>
+    <version>2.1.0-SNAPSHOT</version>
```
I would rather not conflate upgrading the underlying version w/ adding join support. Can that be done in a separate PR, please?

xumingming replied:
I will remove this change.

takidau commented on dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java:
```diff
@@ -0,0 +1,305 @@
+/**
+ * {@code BeamRelNode} to replace a {@code Join} node.
+ *
+ * <p>Support for join can be categorized into 3 cases:
```
We will need to be careful here. This is a perfectly reasonable initial set of supported use cases, but I think there will be more variations we want to support in the future. As long as SQL remains in a branch, we can probably forge ahead with this change, but I will want to make sure we have a comprehensive plan defined for JOIN support before merging to master. Let me try to get my SQL spec doc finished soon, that will hopefully give us a good place to discuss what we want to pursue.

xumingming replied:
Yeah, looking forward to your spec :)

takidau replied:
Thanks for your patience. The spec proposal is not quite ready yet, but close. At this point though, I think what you have should be compatible with what I'm proposing. There will be a separate, but related, question of default triggering, but we can tackle that as part of the merge burndown list (added there).

takidau commented on dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java:
```diff
@@ -0,0 +1,305 @@
+  //       currently joinlibrary does not support FULL OUTER JOIN
+  //       once FULL OUTER JOIN is supported in joinlibrary we will use
+  //       joinlibrary instead.
+  //       Pull Request: https://github.com/apache/beam/pull/3267
```
It looks like this PR is merged, correct? Please resolve this if so.

xumingming replied:
Will do this after rebase from DSL_SQL.

xumingming commented:
Rebase done.

xumingming commented:
Rebase done.

XuMingmin commented on dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedBeamSqlTable.java:
```diff
@@ -49,6 +48,7 @@
   public static final ConcurrentLinkedQueue<BeamSqlRow> CONTENT = new ConcurrentLinkedQueue<>();
 
   private List<BeamSqlRow> inputRecords;
+  private PCollection.IsBounded isBounded = PCollection.IsBounded.BOUNDED;
```
not sure if this can test both `BOUNDED` and `UNBOUNDED` case, as `Create.of()` returns a BOUNDED PCollection.  Could use `TestStream` IMO.

xumingming replied:
The core feature of `TestStream` is enable tests to advance the watermark and processing time as elements are emitted. Which seems not needed for our test? We just need `IsBounded` to tell whether specified tables can join?

XuMingmin replied:
I don't know if you can change the BOUNDED/UNBOUNDED behavior of PCollection by set it here. The test for UNBOUNDED is more about triggers/watermark to me.

xumingming replied:
@XuMingmin make sense. Let me add a `MockedUnboundedTable` and add real test for unbounded table.

XuMingmin commented:
@takidau I'm thinking to finish this PR first as an initial implementation of `JOIN`, to be included in the merge_to_master thread.  We could keep enhance to align with your proposal specs, I'm sure there's much space to improve.

takidau commented:
I'm going to dive into this tomorrow (no more training :-). I basically want to make sure we don't introduce anything we might want to roll back in the future semantics-wise.

xumingming commented on dsls/sql/pom.xml:
```diff
@@ -149,6 +149,11 @@
       <scope>provided</scope>
     </dependency>
     <dependency>
+      <groupId>org.apache.kafka</groupId>
```
Should be removed.

xumingming commented:
Update: 

1. Added a `MockedUnboundedTable` which use `TestStream` to mock a _real_ unbounded table.
1. Updated the JOIN test to use `MockedUnboundedTable`.

xumingming commented:
Since now we have `MockedBeamSqlTable`(might better be renamed to MockedBoundedTable) and `MockedUnboundedTable`, some refactoring is needed to cleanup the mock logic and related tests, created https://issues.apache.org/jira/browse/BEAM-2515 to track it.

xumingming commented:
you mean `kafka-clients`, right? Seems it is actually needed. Our pom has configured the maven dependency plugin to check whether there are unused dependency, and whether there are dependency which are used but not declared, so the code can build, the dependencies should be correct, we seems not need to review the pom manually?

XuMingmin commented:
Make sense, let's keep it.

takidau commented on dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java:
```diff
@@ -0,0 +1,275 @@
+    return nullRow;
+  }
+
+  private List<Pair<Integer, Integer>> extractJoinColumns(int separator) {
```
What is separator here? Is it just an offset used to effectively remove any columns on the left side? If so, it could maybe use a better name.

xumingming replied:
rename to `leftRowColumnCount`.

takidau commented on dsls/sql/src/main/java/org/apache/beam/dsls/sql/transform/BeamJoinTransforms.java:
```diff
@@ -0,0 +1,166 @@
+    private PCollectionView<Map<BeamSqlRow, Iterable<BeamSqlRow>>> sideInputView;
+    private JoinRelType joinType;
+    private BeamSqlRow rightNullRow;
+    private boolean swap;
```
Final for all?

takidau commented on dsls/sql/src/test/java/org/apache/beam/dsls/sql/TestUtils.java:
```diff
@@ -0,0 +1,99 @@
+   */
+  public static class RowsBuilder {
+    private BeamSqlRecordType type;
+    private List<BeamSqlRow> rows = new ArrayList<>();
```
Both final? I think you'll have to add a constructor, but it's the proper way to do it unless type needs to be mutable.

takidau commented on dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedUnboundedTable.java:
```diff
@@ -0,0 +1,120 @@
+ * A mocked unbounded table.
+ */
+public class MockedUnboundedTable extends MockedTable {
+  private List<Pair<Duration, List<BeamSqlRow>>> timestampedRows = new ArrayList<>();
```
final?

xumingming replied:
The `timestampRows` are configured step by step, so can't be set to `final`.

takidau commented on dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java:
```diff
@@ -0,0 +1,275 @@
+    BeamSqlRow realRightNullRow = swapped ? leftNullRow : rightNullRow;
+    JoinRelType realJoinType = swapped ? JoinRelType.LEFT : joinType;
+
+    final PCollectionView<Map<BeamSqlRow, Iterable<BeamSqlRow>>> rowsView = realRightRows
```
Factor everything below here into a helper function that's called with the rows already swapped if necessary? Will make reading the code easier and less error prone.

takidau commented on dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java:
```diff
@@ -0,0 +1,275 @@
+ * <p>For the first two cases, a standard join can be utilized to implement them as long as the
+ * windowFn of the both sides match. For the third case, {@code sideInput} is utilized to implement
+ * the join, hence there are some constrains for the third case: 1) FULL JOIN is not supported
+ * 2) The unbounded table must be at the left side of the OUTER JOIN.
```
The left side constraint is an implementation detail, isn't it? It appears you automatically swap left and right sides internally for the sake of doing the join then swap the results back if the unbounded table is on the right, don't you?

xumingming replied:
Yes, I have done the auto-swap. The javadoc should be updated.

xumingming replied:
Actually there is still a constraint, unbounded table should on the left side of the join after auto-swap, e.g. we don't support the follow join:

```sql
select * from bounded_table LEFT OUTER JOIN unbounded_table.
```

takidau commented on dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java:
```diff
@@ -0,0 +1,275 @@
+ * the join, hence there are some constrains for the third case: 1) FULL JOIN is not supported
+ * 2) The unbounded table must be at the left side of the OUTER JOIN.
+ *
+ * <p>There is also some overall constrains:
```
constraints

takidau commented on dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java:
```diff
@@ -0,0 +1,275 @@
+ *
+ * <ul>
+ *  <li>Only equi-join is supported</li>
+ *  <li>CROSS JOIN is not supported</li>
```
Should there be tests verifying attempts to violate these constraints fail gracefully?

xumingming replied:
Will do.

xumingming commented on dsls/sql/src/main/java/org/apache/beam/dsls/sql/transform/BeamJoinTransforms.java:
```diff
@@ -0,0 +1,166 @@
+  /**
+   * As the method name suggests: combine two rows into one wide row.
+   */
+  private static BeamSqlRow combineTwoRowsIntoOne(BeamSqlRow leftRow,
```
Good point!

xumingming commented on dsls/sql/src/test/java/org/apache/beam/dsls/sql/rel/BeamJoinRelUnboundedVsBoundedTest.java:
```diff
@@ -0,0 +1,216 @@
+  }
+
+  @Test
+  public void testInnerJoin() throws Exception {
```
Good point, find a related bug.

xumingming commented on dsls/sql/src/test/java/org/apache/beam/dsls/sql/rel/BeamJoinRelUnboundedVsUnboundedTest.java:
```diff
@@ -0,0 +1,222 @@
+    // 2, 2 | NULL, NULL
+    // ---- | -----
+    // 2, 2 | 2, 5
+    // 3, 3 | NULL, NULL
```
Yeah, I missed this one, will fix.

xumingming commented on dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedBeamSqlTable.java:
```diff
@@ -119,10 +117,11 @@ public RelDataType apply(RelDataTypeFactory a0) {
   @Override
   public BeamIOType getSourceType() {
-    return BeamIOType.UNBOUNDED;
+    return BeamIOType.BOUNDED;
```
I intended to make this change in another PR(to avoid unrelated changes). Created a separate issue to track this: https://issues.apache.org/jira/browse/BEAM-2515

takidau replied:
sgtm, thx

xumingming commented on dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedTable.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Base class for mocked table.
+ */
+public abstract class MockedTable extends BaseBeamTable {
```
ditto, will do the change in a dedicated PR.

xumingming commented on dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedUnboundedTable.java:
```diff
@@ -0,0 +1,120 @@
+/**
+ * A mocked unbounded table.
+ */
+public class MockedUnboundedTable extends MockedTable {
```
ditto.

xumingming commented on dsls/sql/src/main/java/org/apache/beam/dsls/sql/rule/BeamJoinRule.java:
```diff
@@ -0,0 +1,53 @@
+public class BeamJoinRule extends ConverterRule {
+  public static final BeamJoinRule INSTANCE = new BeamJoinRule();
+  private BeamJoinRule() {
+    super(LogicalJoin.class, Convention.NONE,
```
because the first param of `ConverterRule`(`BeamJoinRule`'s parent) is(I quote):

```
Type of relational expression to consider converting
```

For a join, before convert, it's relation algebra tree is something like:

```bash
LogicalProject(order_id=[$0], sum_site_id=[$1], order_id0=[$2], sum_site_id0=[$3])
  LogicalJoin(condition=[=($0, $2)], joinType=[inner])
    LogicalProject(order_id=[$0], sum_site_id=[$2])
      LogicalAggregate(group=[{0, 1}], sum_site_id=[SUM($2)])
        LogicalProject(order_id=[$0], $f1=[TUMBLE($3, 3600000)], site_id=[$1])
          LogicalTableScan(table=[[ORDER_DETAILS]])
    LogicalProject(order_id=[$0], sum_site_id=[$2])
      LogicalAggregate(group=[{0, 1}], sum_site_id=[SUM($2)])
        LogicalProject(order_id=[$0], $f1=[TUMBLE($3, 3600000)], site_id=[$1])
          LogicalTableScan(table=[[ORDER_DETAILS]])
```

Note the `LogicalJoin` in the tree, it is the class we are trying to convert to `BeamJoinRel`.

takidau replied:
Makes sense, thx for the detailed explanation. 👍

xumingming commented:
Updated the code according to review comments @takidau

takidau commented on dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java:
```diff
@@ -242,17 +266,22 @@ private BeamSqlRow buildNullRow(BeamRelNode relNode) {
-  private List<Pair<Integer, Integer>> extractJoinColumns(int separator) {
+  private List<Pair<Integer, Integer>> extractJoinColumns(int leftRowColumnCount) {
+    // it's a CROSS JOIN because: condition == true
+    if (condition instanceof RexLiteral && (Boolean) ((RexLiteral) condition).getValue()) {
```
For my own edification, can you explain this a little further? What is the condition here, and why does it being set to true imply CROSS JOIN?

xumingming replied:
For the following cross join:

```sql
select * from person p1, person p2
```

its relation algebra tree is:

```
LogicalProject(id=[$0], name=[$1], age=[$2], ts=[$3], boolean_c=[$4], id0=[$5], name0=[$6], age0=[$7], ts0=[$8], boolean_c0=[$9])
  LogicalJoin(condition=[true], joinType=[inner])
    EnumerableTableScan(table=[[person]])
    EnumerableTableScan(table=[[person]])
```

so in Calcite, a `CROSS JOIN` is converted to a `INNER JOIN` with the `condition` set to `true`.

takidau commented:
Merged.

xumingming commented:
Thanks @takidau @XuMingmin
